src/react/elements.js

```diff
@@ -33,14 +33,18 @@ function createPropsObject(
   let ref = realm.intrinsics.null;
 
   const setProp = (name: string, value: Value): void => {
-    if (name === "key" && value !== realm.intrinsics.null) {
+    if (name === "key" && value !== realm.intrinsics.null && value !== realm.intrinsics.undefined) {
       key = computeBinary(realm, "+", realm.intrinsics.emptyString, value);
-    } else if (name === "ref") {
+    } else if (name === "ref" && value !== realm.intrinsics.undefined) {
       ref = value;
     } else if (name !== "__self" && name !== "__source") {
       invariant(props instanceof ObjectValue);
       Properties.Set(realm, props, name, value, true);
     }
+    if (key === "children" && (value instanceof ObjectValue || value instanceof AbstractObjectValue)) {
```

Comment:
Children passed in should not be marked as final. Only children arrays created by React.